### PR TITLE
add junit recording in e2e runner

### PR DESCRIPTION
**Pull request by freehan**
2nd part of #27825  

Injected junit recorder at a few places.

### CODE REVIEW

File: hack/jenkins/e2e-runner.sh [hunk 1]
```diff
@@ -21,8 +21,53 @@ set -o nounset
 set -o pipefail
 set -o xtrace
 
+# include shell2junit library
+source <(curl -fsS --retry 3 'https://raw.githubusercontent.com/kubernetes/kubernetes/master/third_party/forked/shell2junit/sh2ju.sh')
```
Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: Why do you need to curl this? Shouldn't this file be included in whatever PR is getting tested?

Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: IMO it's not good to get it from master-- if it breaks at head it breaks everywhere...

File: hack/jenkins/e2e-runner.sh [hunk 2]
```diff
@@ -391,9 +433,13 @@ fi
...
 if [[ "${E2E_UP:-}" == "${E2E_DOWN:-}" && -f "${gcp_resources_before}" && -f "${gcp_resources_after}" ]]; then
   difference=$(diff -sw -U0 -F'^\[.*\]$' "${gcp_resources_before}" "${gcp_resources_after}") || true
+  noleak=true
   if [[ -n $(echo "${difference}" | tail -n +3 | grep -E "^\+") ]] && [[ "${FAIL_ON_GCP_RESOURCE_LEAK:-}" == "true" ]]; then
+    noleak=false
+  fi
+  if ! ${noleak} ; then
     echo "${difference}"
     echo "!!! FAIL: Google Cloud Platform resources leaked while running tests!"
-    exit 1
```
Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: Does the presence of the junit file cause a failure now? Or is an exit 1 still needed somewhere?

File: hack/jenkins/e2e-runner.sh (hunk 1 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: The reason we need to curl it is because e2e-runner is running as standalone script on jenkins. It does not come with the whole repo.

File: hack/jenkins/e2e-runner.sh (hunk 2 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: we do not need to exit here because the record_command will exit upon failure. `false` is treated as failure.

File: hack/jenkins/e2e-runner.sh [hunk 3]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
 
+# record_command records the command output and error message in junit format
+# it expects the first argument to be the class and the second to be the name of the command
+# Example:
+# record_command PRESETUP curltest curl google.com
+# record_command CLEANUP check false
+#
+# WARNING: variable changes will NOT be effective after record_command returns
+function record_command() {
+    set +o xtrace
```
Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: @david-mcmahon you seem like someone who knows bash. Would you like to review this?

**Comment by freehan** [comment]
fyi, the junit result showed up.

https://00e9e64bacaaab0343c9f68610ca562f511526f3e493ed4d10-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F28133%2Fkubernetes-pull-build-test-e2e-gce%2F46972%2Fartifacts%2FTEST-CLEANUP.xml?qk=AD5uMEuoe9yH03gUZlnHm7j_DI4J9QMJSXt16_OjhnxI0hXnoZBzq7gJ3kB0jrnobvXvcqr2WKbPgo9FHfDvDnkuyMptlyMqdm-In-eaU60T7tefakJm-6HCGTeiENOZdLVxpc3p76X-LU1qyUwqbitJR7Z2asTWIoTyQzkkEekmSicAIkSzGETdNWKjQe9KJl5XrzoLVeOgSHDFnRO14vTB8vUwtEo6I1FuE1LpLTB7tq7UZrUAuqy9XbMpI0aB8AExtErMz5Vhb-5DbqKbUOEX4S5r-pRC_JpjWACBahw57zi0iuirp3wj-Pe8nm4k6KwEUh1TsyhX6cyByjTck3Sh6d_XME4KOYPAj0jVrOZ-AHP8H08bwOzFRxYdQ3Rocsk3xXckyy2DAWYtdl_w1n5IjmWMPT4zF7NXdxuU5DzmaYRt8_1Wfvcl_EkRvD7bCRy0n1O7BbML3RiLBuoyXIaY-meY-i_88jZy19VemduHfTw9HJyJVTFdeLEVyqpLimCqjgJI3twqzJpqv3dKiz_7NfZxcjKSzNz03XA8akboYvWaDEqh3Yh73K6H3ZdoMYCsGvI-40i4EwNrWY6O9MBt8w_YrnpQLkPdCX49afG2mWaGJvEoR7uuWZryk_Rg_6xlF05nwRyRn-roBDhEw2O5sgwtWIRcEG0BDL82y9RFdAN7zMkRAA3NgJ3O4unIlfeZgWHtWuZWTZI6BgyaieY1W-9T3hfNvNUkt74PKiwo_L0EMuBQdZZ1L8_KrQ4mFpyYxVWU6Afxk9cSV1GkzeLC-fKrOX2vFhTt5ckiVbA04LzhXQ8nIT7P2l034fwdecbfIq2MvaSXzYnNmkByJ8G9wybZfPk2WZPDNskGrgpiOo6f7yZ0SRlraWAB-sw1zdBM2pKzyyMOCapp2VB4HftVfD9fRIY0Ea2aaJotEyJKGp8-BPn_JLp4XXBb1jNrMdsqhDrer-a5

File: hack/jenkins/e2e-runner.sh (hunk 1 above)
Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: How is this running standalone? Don't you need to clone this repo to get access to the script in the first place?

File: hack/jenkins/e2e-runner.sh [hunk 4]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+    set +o nounset
+    set +o errexit
```
Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: Why set + here and then re-set - below?

File: hack/jenkins/e2e-runner.sh [hunk 5]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+    echo "Running command: $@"
+    local -r ret=$(juLog -output="${ARTIFACTS}" -class="${class}" -name="${name}" "$@")
+
+    set -o nounset
+    set -o errexit
+    set -o xtrace
+
+    # since record_command disable errexit, exit script on failure.
+    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
+    if [ ${found} -le 0 ]; then
```
Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: Does juLog return 0 or 1? Why not just:
if juLog ...; then
fi

File: hack/jenkins/e2e-runner.sh (hunk 4 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: for xtrace, if we leave it on, it will trace all commands in juLog, which is way too noisy. 

juLog made the assumption that nounset and errexit were not set. I want to ensure it won't exit accidentally.

File: hack/jenkins/e2e-runner.sh (hunk 1 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: Nope. e2e-runner will pull k8s binary from gcs. 
@ixdy

File: hack/jenkins/e2e-runner.sh (hunk 5 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: will take a look at this

Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: juLog returns a bunch of stuff. I have to ensure the output indicates the command is successful.

Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: This PR LGTM when bash is good :)

File: hack/jenkins/e2e-runner.sh (hunk 2 above)
Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: I typically like to return control to main and exit from there, but it's a style choice.

Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: errexit is set, hence e2e runner may exit anywhere.

File: hack/jenkins/e2e-runner.sh (hunk 5 above)
Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: Looks fine to me.

File: hack/jenkins/e2e-runner.sh [hunk 6]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+# record_command PRESETUP curltest curl google.com
+# record_command CLEANUP check false
+#
+# WARNING: variable changes will NOT be effective after record_command returns
+function record_command() {
+    set +o xtrace
+    set +o nounset
+    set +o errexit
+
+    class=$1
```
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: slight preference for adding `local` keyword so these don't leak outside the function

Comment (david-mcmahon) [inline hack/jenkins/e2e-runner.sh]: Thanks.  I didn't see that.  High preference for that.

File: hack/jenkins/e2e-runner.sh [hunk 7]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+    shift
+    name=$1
+    shift
+    echo "Recording: ${class} ${name}"
+    echo "Running command: $@"
```
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: I think you could use `${@:3}` instead of `shift`ing, though I haven' tested this.

File: hack/jenkins/e2e-runner.sh (hunk 5 above)
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: style nit: prefer `[[ ]]` over `[ ]`

File: hack/jenkins/e2e-runner.sh [hunk 8]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+
 : ${KUBE_GCS_RELEASE_BUCKET:="kubernetes-release"}
 
+# record_command records the command output and error message in junit format
+# it expects the first argument to be the class and the second to be the name of the command
+# Example:
+# record_command PRESETUP curltest curl google.com
+# record_command CLEANUP check false
+#
+# WARNING: variable changes will NOT be effective after record_command returns
```
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: what does this mean?

**Comment by ixdy** [comment]
I think our tooling looks for junit files named `junit*.xml`, so maybe we want to change the names?

File: hack/jenkins/e2e-runner.sh (hunk 7 above)
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: it's also pretty readable as-is, so maybe we don't need to worry about this.

File: hack/jenkins/e2e-runner.sh (hunk 6 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: good catch. Will fix it.

File: hack/jenkins/e2e-runner.sh (hunk 5 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: sure

**Comment by freehan** [comment]
Oh yes. Will change that.

File: hack/jenkins/e2e-runner.sh (hunk 8 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: for instance:
var=1
record_command "Class" "Name" var=2

var will still be 1, because juLog will execute the actual command in a subshell. 

After it returns, Var will still be 1 because everything got executed in a subshell.

File: hack/jenkins/e2e-runner.sh (hunk 7 above)
Comment (freehan) [inline hack/jenkins/e2e-runner.sh]: it works. Thanks.

File: hack/jenkins/e2e-runner.sh (hunk 8 above)
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: ah, maybe rephrase slightly to "variable changes *in the command* will NOT ..."

File: hack/jenkins/e2e-runner.sh [hunk 9]
```diff
@@ -21,8 +21,53 @@ set -o nounset
...
+
+# E2E runner stages
+STAGE_PRE="PRE-SETUP"
+STAGE_SETUP="SETUP"
+STAGE_CLEANUP="CLEANUP"
+STAGE_KUBEMARK="KUBEMARK"
+
 : ${KUBE_GCS_RELEASE_BUCKET:="kubernetes-release"}
 
+# record_command records the command output and error message in junit format
```
Comment (ixdy) [inline hack/jenkins/e2e-runner.sh]: it's not super clear to me from the name/comments that this actually runs the command too,

File: hack/jenkins/e2e-runner.sh [hunk 10]
```diff
@@ -21,8 +21,54 @@ set -o nounset
...
+
+    set -o nounset
+    set -o errexit
+    set -o xtrace
+
+    # since record_command disable errexit, exit script on failure.
+    local -r found=$(echo "${ret}" | grep -c "+++ exit code: 0")
+    if [[ ${found} -le 0 ]]; then
+        echo "Last command failed, exiting..."
+        exit 1
```
Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: The goal here is to get a build marked "unstable" instead of "failed" which I understand means, now that I unfortunately learned about jenkins, that we shouldn't exit at all (we want to run all the commands) and should print "Test Suite Failed" so that jenkins will mark it unstable (@ixdy is about to turn that plugin on). @ixdy, sound correct?

Comment (lavalamp) [inline hack/jenkins/e2e-runner.sh]: Hold on that. @fejta is trying to convince us that we shouldn't care about
the difference and that it's perfectly valid to fail like this.

**Comment by freehan** [comment]
ping

**Comment by lavalamp** [comment]
Whoops, sorry, didn't realize this was waiting on me. LGTM

File: hack/jenkins/e2e-runner.sh (hunk 1 above)
Comment (spxtr) [inline hack/jenkins/e2e-runner.sh]: @freehan is correct. We curl down `e2e-runner.sh` which downloads the kubernetes tar from GCS.

**Comment by spxtr** [comment]
```
    <testsuite failures="0" assertions="" name="CLEANUP" tests="1" errors="0" time="">
    
    <testcase assertions="1" name="gcp_resource_leak_check" time="" classname="CLEANUP">
    
    <system-out>
<![CDATA[

+++ Running case: CLEANUP.gcp_resource_leak_check 
+++ working dir: /var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/kubernetes
+++ command: true
+++ exit code: 0
]]>
    </system-out>
    <system-err>
<![CDATA[

]]>
    </system-err>
    </testcase>
  
    </testsuite>
```

That's pretty strange looking. I suppose it's valid.
